### PR TITLE
JS: Fix comments inside template

### LINKED ISSUE #9274: Inline comment issue with `html` tagged template
If I use a template tag called `html`, Prettier moves my inline comment outside of the curly braces. This does not happen if I change `html` to another function name.

---

**Prettier 2.1.2**
[Playground link](https://prettier.io/playground/#N4Igxg9gdgLgprEAuEAnOMCuqoAIAWMAtgDYAGAOnrgDwAmAlgG4B8VuHuAJMDPgwGcAdAIy4AZONx9BIjEICGqGIJjtOAegBUuBUwgM6uSFAEQScXHFSoIqAbgYAzXGRnDRMMo4eYodOCcGKDgjLQ0AX3UaDUZWKjIAbioQABoQCAAHFWgBZFAlWwB3AAUlBDyUBRIihQBPPPSAI1QFMABrDABlTLbggHNkGFRMOHS4IibQgLoAGQUofswFfrgAMTsiBRgVReQQBUwYCDSQQlIAdX54AV6wOC6KhhUmZ7r9sAFGkGDRZRLWv0tsgnNVROkAFYCAAeACFWh1ugoiHBZsE4CCwWMQFDoV0BhYAIqYCDwTEkcEgXr2az7JoKKYkU6ZVDBGAXQx8ZAADgADOkWRBRBdWpl9iy4H8mBj0gBHEnwAFZSoHAQAWhCoVCp3Q8oY6ABK2BSFBFOxoiIDCGI3NBLgxNJGJNWPSMAZHLoXKQACZXa0GCQBgBhCBEY0gSUAVlOmFEABUGZVTZSmKMAJL+BAwLpgVnZACC-i6MDqFnJogiESAA)
```sh
--parser babel
```

**Input:**
```js
return html`
  <div>
    ${this.set && this.set.artist
    /* avoid console errors if `this.set` is undefined */}
  </div>
`;

```

**Output:**
```js
return html`
  <div>
    ${this.set && this.set.artist}
    /* avoid console errors if `this.set` is undefined */
  </div>
`;

```

**Expected behavior:**
```js
return html`
  <div>
    ${
      this.set && this.set.artist
      /* avoid console errors if `this.set` is undefined */
    }
  </div>
`;

### PR DESCRIPTION
**Pull request by fisker**
Fixes #9274

- [x] I’ve added tests to confirm my change works.
- [ ] (If changing the API or CLI) I’ve documented the changes I’ve made (in the `docs/` directory)
- [x] (If the change is user-facing) I’ve added my changes to `changelog_unreleased/*/pr-XXXX.md` file following `changelog_unreleased/TEMPLATE.md`.
- [x] I’ve read the [contributing guidelines](https://github.com/prettier/prettier/blob/master/CONTRIBUTING.md).

**✨[Try the playground for this PR](https://prettier.io/playground-redirect)✨**

### CODE REVIEW

File: changelog_unreleased/javascript/pr-9278.md [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+graphql`
+  ${foo
+  /* comment */
+  }
+`;
+css`
+  ${foo
+  /* comment */
+  }
+`;
```
Comment (fisker) [inline changelog_unreleased/javascript/pr-9278.md:60]: This is not very pretty, but I'm not sure how to improve it.

Comment (fisker) [reply changelog_unreleased/javascript/pr-9278.md]: Improved style https://github.com/prettier/prettier/pull/9278/commits/a35637f4d4e95217966365bb6bbd6361e6cfca59, but feel there is something wrong.

**Review by evilebottnawi — APPROVED** [review]
Looks better

**Comment by thorn0** [comment]
I think you need to add `lineSuffixBoundary` like it's done in `printer-estree` when printing usual template literals.

**Comment by thorn0** [comment]
Ideally, it'd be good to extract the code from `printer-estree`/`case TemplateLiteral` to a function to reuse it for printing these expression. In the end, there should be no difference in how expressions are formatted in usual and multiparser template literals:

**Prettier pr-9278**
[Playground link](https://deploy-preview-9278--prettier.netlify.app/playground/#N4Igxg9gdgLgprEAuEcAeAHATgRgAQC8eAFjALYA2ABgDpR54A8AJgJYBuAfACTBoAUdAGYQIAGjwB6SXiEBzOgwBGAQywBKAL6NJbLnSoBuOnXTYATITy16TPTz6CoI8VJnzFeVRu26OnA2MoEDEQCAwYVmgAZ2RQNSwIAHcABTUEWJQVCiSVAE9Y0KUsFTAAazgYAGUMUtYoOWQYLABXOFC4MiU4ZmYegBkVBpaVOTgAMQgsMhUYSIbkEBUWmAgQkFJKAHViVnho2rA4Koy9jj28xbBowpB66LgsGBSSuRnkIWyH0IAraLQAEIlcqVKoqMhwfr1OAfL7tEB-NBVepyChwACKLQg8FhFG+IFqWAeWEWqm6FHW2HqMC2rGYMGIyAAHAAGULYCAPLYlDCLbBwYnsGGhACOWPgL3CmSW0QAtFA4D0eussHAxaxVS9Ru8kJ88fCHmRWE1WgaUWjMdiYbq4aEYColLT6YykOY7SVWBQUQBhCBkHWoaIAVnWLQeABUHZk9fj2G0AJJQPqwKpgLCsCIAQSTVRgeTRuIemk0QA)
```sh
--parser babel
```

**Input:**
```jsx
expr1 = html`
  <div>${x(
foo, // fg
  bar)}</div>
`;

expr2 = `
  <div>${x(
foo, // fg
  bar)}</div>
`;

```

**Output:**
```jsx
expr1 = html`
  <div>
    ${
      x(
        foo, // fg
        bar
      )
    }
  </div>
`;

expr2 = `
  <div>${x(
    foo, // fg
    bar
  )}</div>
`;

```

**Comment by fisker** [comment]
> I think you need to add `lineSuffixBoundary` like it's done in `printer-estree` when printing usual template literals.

I tried that, https://github.com/prettier/prettier/pull/9278/commits/a93fd70cfbe5d9e5d88ec0f677d67d53cacc53b9

Looks not very pretty.

**Comment by thorn0** [comment]
I see. I looked deeper into this. Without doing what I wrote in my [previous comment](https://github.com/prettier/prettier/pull/9278#issuecomment-699935011), `printTemplateExpressionDoc` can be rewritten this way, without `findInDoc`:

```js
function printTemplateExpressionDoc(doc) {
  return group(
    concat([
      "${",
      indent(concat([softline, doc])),
      softline,
      lineSuffixBoundary,
      "}",
    ])
  );
}
```

`lineSuffixBoundary` doesn't seem to make difference because line breaks are added by the code that prints comments, but I thought it's better to keep it to stay on the safe side.

**Comment by fisker** [comment]
@thorn0 This changes lot https://github.com/prettier/prettier/pull/9278/commits/83535c5ce35080c16c633382426e2016304a4702

**Comment by thorn0** [comment]
Not good, unmotivated changes should be avoided. Let's switch back to your implementation.

I don't like that it reacts to comments in the middle of the expression, e.g. here:
```js
expr1 = html`
  <div>
    ${
      x(
        foo, // fg
        bar
      )
    }
  </div>
`;
```
But I'm not sure what can be done about that.

**Comment by fisker** [comment]
> I don't like that it reacts to comments in the middle of the expression

Yes, that's the problem

File: src/language-js/embed.js [hunk 2]
```diff
@@ -655,4 +657,15 @@ function printHtmlTemplateLiteral(path, print, textToDoc, parser, options) {
...
 }
 
+function printTemplateExpression(print) {
+  return (path) => {
+    const node = path.getValue();
+    let printed = print(path);
+    if (node.comments && node.comments.length) {
+      printed = concat([indent(concat([softline, printed])), softline]);
+    }
+    return concat(["${", printed, lineSuffixBoundary, "}"]);
```
Comment (thorn0) [inline src/language-js/embed.js:667]: I'm not totally sure that it's impossible for a descendant node's comment to get printed after the expression. There will be no indentation in that at case, but at least `lineSuffixBoundary` will prevent it from being printed after the closing brace.

**Comment by fisker** [comment]
Looks good

**Comment by thorn0** [comment]
An example of a situation where `lineSuffixBoundary` helps. It's unstable (because of #7746 or something similar), but at least the code doesn't get broken.

**Prettier pr-9278**
[Playground link](https://deploy-preview-9278--prettier.netlify.app/playground/#N4Igxg9gdgLgprEAuEALGBbANgAwDwAmAlgG4B8AJMAIYD8ARkgBRgD0rYAOlAJQC+eVsXI4QAGhAQADjCLQAzslDUATiogB3AAqqEilNSwbqAT0UT6K6mADWcGAGUp1olADmyGCoCucCXAx6OAICYIAZandvajc4ADEIFQxqGFl3ZBBqbxgIcTRMLAB1VCJ4eWcwOAc9UtJSkwyweXMQV3k4FRgtKzdk5AAzQ3aJACt5AA8AIStbewdqDDgw1zgBob8QMfGHVzcsOABFbwh4NaxhkGcVdpUM+mogrDypFVcYQqICGFRkAA4ABgkLwg7UKVikGRecBuJFWEgAjsd4N1pPpMvIALRQODBYJ5FRwRFEAndGJ9JCDc4bdoYIieHzU3b7I4nVYU9YSGAPD5fH5IABMnKsRCwuwAwhAMOSQNCAKx5bztAAqD30lIuJF8AEkoKFYA4wK8ZABBXUOGAmfZndp8PhAA)
```sh
--parser babel
```

**Input:**
```jsx
html`<div>${a?b:(c//c
)}</div>`
```

**Output:**
```jsx
html`<div>
  ${a
    ? b
    : c //c
  }
</div>`;

```

**Second Output:**
```jsx
html`<div>
  ${
    a ? b : c //c
  }
</div>`;

```

**Review by thorn0 — APPROVED** [review]

**Review by sosukesuzuki — APPROVED** [review]

**Review by kadirselcuk — APPROVED** [review]

**Review by kadirselcuk — APPROVED** [review]

**Pull request merged**